src/config/resolver.rs

```diff
@@ -0,0 +1,96 @@
+use std::collections::VecDeque;
+
+use crate::config::is_default;
+use crate::config::{Config, Source};
+use anyhow::anyhow;
+use serde::{Deserialize, Serialize};
+use tokio::fs::File;
+use tokio::io::AsyncReadExt;
+use url::Url;
+
+#[derive(Default, Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+enum LinkType {
+  #[default]
+  Config,
+  GraphQL,
+  Protobuf,
+  Data,
+}
+
+#[derive(Default, Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub struct Link {
+  #[serde(default, skip_serializing_if = "is_default")]
+  type_of: LinkType, // Type of the link
+  #[serde(default, skip_serializing_if = "is_default")]
+  src: String, // Source URL for linked files
+  #[serde(default, skip_serializing_if = "is_default")]
+  id: Option<String>, // Id is used to refer at different places in the config
+  #[serde(default, skip_serializing_if = "is_default")]
+  content: Option<String>, // Stores raw content
+}
+
+impl Link {
+  pub async fn resolve_recurse(self) -> anyhow::Result<Vec<Link>> {
+    let mut result: Vec<Link> = Vec::new();
+    if self.type_of == LinkType::Protobuf || self.type_of == LinkType::Data {
+      let link: Link = Self::get_raw_content(&self).await?;
+      result.push(link);
+      return Ok(result);
+    }
```

Comment:
since all files can be both `http://` or `file://`, why we handle `Protobuf` and `Data` differently?

Reply:
Since we don't need to detect the source type for these files. Because Config and GraphQL files could have recursive links, but these files don't. To, remove unnecessary processing I have created a different function for these two types

Reply:
from my more preferred to least preferred, I'd say that we have a few alternatives:

1. load source every time and don't check for it at all, as it'll just not find a link to recursive resolve;
2. move this condition to lines 48 and 63, early returning in case it matches;
3. move the loading (of both file/network protocol) to specific functions that can be re-used in both cases;

in case we go with `2.`, feel free to use `get_raw_content` as much as we don't duplicate the code for loading stuff (so this method will need to rely on the new methods).

IMO: if the processing is not high costly, we can certainly load everything and early return in case is `Protobuf` or `Data`.